[optim] update step in functional and pass state_steps instead of state

mikaylagawarecki opened the pull request:
Updated 
- Adagrad
- Adamax
- Adam
- AdamW
- RAdam
make multi_tensor functionals take `state_steps: List[Tensor]` instead of taking `states: List[Dict]`
make `state_steps: List[int]s -> state_steps:List[Tensor]` where each is a Singleton tensor so step can be updated within the functional

(NAdam and ASGD) were updated in separate diffs to fold their handling of state into the functionals

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* #69981
* #69980
* #71335
* #71334
* __->__ #71333

Differential Revision: [D33767872](https://our.internmc.facebook.com/intern/diff/D33767872)

mikaylagawarecki committed:
[optim] update step in functional and pass state_steps instead of state

[ghstack-poisoned]

albanD reviewed (approved):
Nice cleanup!
FYI @mrshenli @ngimel

mikaylagawarecki committed:
Update on "[optim] update step in functional and pass state_steps instead of state"

Updated 
- Adagrad
- Adamax
- Adam
- AdamW
- RAdam
make multi_tensor functionals take `state_steps: List[Tensor]` instead of taking `states: List[Dict]`
make `state_steps: List[int]s -> state_steps:List[Tensor]` where each is a Singleton tensor so step can be updated within the functional

(NAdam and ASGD) were updated in separate diffs to fold their handling of state into the functionals




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767872).

mikaylagawarecki committed:
Update on "[optim] update step in functional and pass state_steps instead of state"

Updated 
- Adagrad
- Adamax
- Adam
- AdamW
- RAdam
make multi_tensor functionals take `state_steps: List[Tensor]` instead of taking `states: List[Dict]`
make `state_steps: List[int]s -> state_steps:List[Tensor]` where each is a Singleton tensor so step can be updated within the functional

(NAdam and ASGD) were updated in separate diffs to fold their handling of state into the functionals


Differential Revision: [D33767872](https://our.internmc.facebook.com/intern/diff/D33767872)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767872).
